BParkingNano/plugins/TrackMerger.cc

```diff
@@ -151,6 +166,61 @@ void TrackMerger::produce(edm::StreamID, edm::Event &evt, edm::EventSetup const
     float DCASig = (DCABSErr != 0 && float(DCABSErr) == DCABSErr) ? fabs(DCABS/DCABSErr) : -1;
     if (DCASig >  dcaSig_  && dcaSig_ >0) continue;
 
+    // clean tracks wrt to all muons
+    int matchedToMuon       = 0;
+    int matchedToLooseMuon  = 0;
+    int matchedToSoftMuon   = 0;
+    int matchedToMediumMuon = 0;
+    for (const pat::Muon &imutmp : *muons) {
+        for (unsigned int i = 0; i < imutmp.numberOfSourceCandidatePtrs(); ++i) {
+            
+            const edm::Ptr<reco::Candidate> & source = imutmp.sourceCandidatePtr(i);
+            if (! ((imutmp.sourceCandidatePtr(i)).isNonnull() && 
+                   (imutmp.sourceCandidatePtr(i)).isAvailable())
+               )   continue;
+
+            const reco::Candidate & cand = *(source);
+            if (cand.charge() == 0 || cand.bestTrack() == nullptr) continue;
+            try { cand.bestTrack()->eta(); }
+            catch(...) {continue;}
+            
+            if (deltaR(trk.eta(),trk.phi(),cand.bestTrack()->eta(),cand.bestTrack()->phi()) < 1.E-5){
```

Comment:
just wondering why we don't simply check that the track (in terms of pair<edm::ProductID,int>) is the same as that of the muon (source.id(), source.key())

An example I did (but needs some changes because the logic of the code was a bit different):
- to save the muon ref https://github.com/ICBPHCMS/cmssw/blob/NanoAOD102X_BPH_10214/PhysicsTools/NanoAOD/plugins/BToKstllProducer.cc#L1040-L1047

- to xcheck the track https://github.com/ICBPHCMS/cmssw/blob/NanoAOD102X_BPH_10214/PhysicsTools/NanoAOD/plugins/BToKstllProducer.cc#L1070-L1077

and I remember that this cleaned a lot the tracks, (I used no ID... and it was much more that 0.3%, now that I think about this...)

Reply:
thanks for the suggestion!
I tried also this option [1] and, provided I implemented it correctly, I get the same results.

![checkMatchingStrategy_Mu](https://user-images.githubusercontent.com/5147955/63718367-b3221080-c84a-11e9-9ec6-9cf1c99fc434.png)

![checkMatchingStrategy_Ele](https://user-images.githubusercontent.com/5147955/63718350-a9001200-c84a-11e9-8eed-35ccd81c4273.png)

[1] https://github.com/sarafiorendi/BParkingNANO/compare/flag_tracks...sarafiorendi:flag_tracks_v2?expand=1

Reply:
I see, thanks for the check. If there is no speed gain in using .id(), .key() nor any other gain, then it's fine as it is
...in the end a dR < 10^-5 is probably satisfied for an object only with itself...

Reply:
actually I was thinking to use the implementation you propose, it looks more clear and we avoid a dR. I'm going to update the PR with your comments this morning